src/Plutarch/Extra/Map.hs

```diff
@@ -1,185 +1,266 @@
-        Term s' (k :--> PBuiltinPair (PAsData k) (PAsData v) :--> PBool)
+        Term
+            s'
+            ( PBuiltinList (PBuiltinPair (PAsData k) (PAsData a))
+                :--> PBuiltinList (PBuiltinPair (PAsData k) (PAsData b))
+                :--> PBool
+            )
     go = phoistAcyclic $
-        plam $ \x p -> unTermCont $ do
-            k' <- pletC (pfromData $ pfstBuiltin # p)
-            pure $ k' #== x
+        pfix #$ plam $ \self ell ell' ->
+            pmatch (puncons # ell) $ \case
+                PNothing -> pmatch (puncons # ell') $ \case
+                    PNothing -> pcon PTrue -- no mismatches found
+                    PJust _ -> pcon PFalse -- one argument too long
+                PJust kv -> pmatch (puncons # ell') $ \case
+                    PNothing -> pcon PFalse -- one argument too long
+                    PJust kv' -> pmatch kv $ \case
+                        PPair h t -> pmatch kv' $ \case
+                            PPair h' t' ->
+                                pif
+                                    ((pkvPairKey # h) #== (pkvPairKey # h'))
+                                    (self # t # t') -- continue
+                                    (pcon PFalse) -- key mismatch
+
+{- | Get the key of a key-value pair.
+
+ @since 1.1.0
+-}
+pkvPairKey ::
+    forall (k :: S -> Type) (v :: S -> Type) (s :: S).
+    (PIsData k) =>
+    Term s (PBuiltinPair (PAsData k) (PAsData v) :--> k)
+pkvPairKey = phoistAcyclic $ plam $ \kv -> pfromData (pfstBuiltin # kv)
+
+{- | Get the value of a key-value pair.
+
+ @since 3.2.0
+-}
+pkvPairValue ::
+    forall (k :: S -> Type) (v :: S -> Type) (s :: S).
+    (PIsData v) =>
+    Term s (PBuiltinPair (PAsData k) (PAsData v) :--> v)
+pkvPairValue = phoistAcyclic $ plam $ \kv -> pfromData (psndBuiltin # kv)
 
--- | @since 1.0.0
-plookup' ::
+{- | Compare two key-value pairs by their keys. Gives 'PTrue' if the key of the
+ first argument pair is less than the key of the second argument pair.
+
+ @since 3.2.0
+-}
+pkvPairLt ::
+    forall (k :: S -> Type) (v :: S -> Type) (s :: S).
+    (PIsData k, PPartialOrd k) =>
+    Term
+        s
+        ( PBuiltinPair (PAsData k) (PAsData v)
+            :--> PBuiltinPair (PAsData k) (PAsData v)
+            :--> PBool
+        )
+pkvPairLt = phoistAcyclic $
+    plam $ \kv kv' ->
+        (pkvPairKey # kv) #< (pkvPairKey # kv')
+
+{- | As 'plookup', but errors when the key is missing.
+
+ @since 3.2.0
+-}
+plookupPartial ::
     forall (k :: S -> Type) (v :: S -> Type) (keys :: KeyGuarantees) (s :: S).
-    (PIsData v, PIsData k, PEq k) =>
+    (PIsData k, PIsData v) =>
     Term s (k :--> PMap keys k v :--> v)
-plookup' = phoistAcyclic $
-    plam $ \x m -> unTermCont $ do
-        res <- pmatchC (plookup # x # m)
-        pure $ case res of
-            PNothing -> ptraceError "plookup': Could not find key."
+plookupPartial = phoistAcyclic $
+    plam $ \k kvs ->
+        pmatch (plookup # k # kvs) $ \case
```

Comment:
we have `passertPJust` on `Plutarch.Extra.Maybe`. I'm perfectly fine with either way, but just letting you know there is a shortcut :D